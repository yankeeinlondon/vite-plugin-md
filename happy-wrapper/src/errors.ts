import { inspect } from './diagnostics'
import { isHappyWrapperError, isInspectionTuple } from './type-guards'

export class HappyMishap extends Error {
  public name = 'HappyWrapper'
  public readonly kind: 'HappyWrapper' = 'HappyWrapper'
  public trace: string[] = []
  constructor(
    message: string,
    options: {
      error?: unknown
      inspect?: unknown
      name?: string
    } = {}) {
    super()
    this.message = `\n${message}`
    if (options.name)
      this.name = `HappyWrapper::${options.name}`

    // proxy if already a HappyWrapper
    if (isHappyWrapperError(options.error))
      this.name = `HappyWrapper::${options.name || options.error.name}`

    if (options.error) {
      const name = options.error instanceof Error
        ? options.error.name.replace('HappyWrapper::', '')
        : 'unknown'
      const underlying = `\n\nThe underlying error message [${name}] was:\n${options.error instanceof Error ? options.error.message : String(options.error)}`
      this.message = `${this.message}${underlying}`
      this.trace = [...this.trace, name]
    }
    else {
      if (options.inspect) {
        const inspections = isInspectionTuple(options.inspect)
          ? [options.inspect]
          : Array.isArray(options.inspect)
            ? options.inspect
            : [options.inspect]

        inspections.forEach((i, idx) => {
          const intro = isInspectionTuple(i) ? `${i[0]}\n` : `${[idx]}:\n`
          const container = isInspectionTuple(i) ? i[1] : i

          this.message = `${this.message}\n\n${intro}${JSON.stringify(inspect(container), null, 2)}`
        })
      }
      if (this.trace.length > 1)
        this.message = `${this.message}\n\nTrace:${this.trace.map((i, idx) => `${idx}. ${i}`)}`
    }
  }
}
