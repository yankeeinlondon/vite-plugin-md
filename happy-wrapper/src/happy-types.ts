import type { Document, DocumentFragment, IElement, INode, IText } from 'happy-dom'
import type { getNodeType } from './utils'
export type InspectionTuple = [msg: string, item: unknown]

export type HTML = string

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type GetAttribute<T extends string> = <N extends Container | HTML>(node: N) => string

export type NodeTypeInput<T extends ReturnType<typeof getNodeType>> = T extends 'html'
  ? string
  : T extends 'text'
    ? IText
    : T extends 'element'
      ? IElement
      : T extends 'document'
        ? Document
        : T

export type NodeType = 'html' | 'text' | 'element' | 'node' | 'document' | 'fragment'
export type NodeSolverInput<T extends NodeType> = T extends 'html'
  ? HTML
  : T extends 'text'
    ? IText
    : T extends 'element'
      ? IElement
      : T extends 'node'
        ? INode
        : T extends 'document'
          ? Document
          : T extends 'fragment'
            ? DocumentFragment
            : unknown

export type DocRoot = Document | DocumentFragment
export type DomNode = IElement | IText | INode
export type Container = DocRoot | DomNode
export type ContainerOrHtml = Container | HTML

export interface ToHtmlOptions {
  pretty?: boolean
  indent?: number
}

export type FragWrapper = DocumentFragment | HTML
export interface BeforeAfterWrapper {
  /** put on the immediate interior of a tag */
  open?: string
  /**
   * a text node _prepended_ before the element
   */
  before?: string
  /** put before the closing tag */
  close?: string
  /**
   * A text node _appended_ after the closing tag
   */
  after?: string
  /**
   * Provides tab/space indentation based on the nesting level
   * of the element
   */
  indent?: number
}

export interface TreeSummary {
  node: string
  children: TreeSummary[]
}

export interface Tree {
  node: Container
  type: string
  level: number
  summary: () => TreeSummary
  toString: () => string

  children: Tree[]
}

/**
 * A callback which receives a node type `C` and allows side-effects and/or mutation
 */
export type UpdateCallback<C extends Container> = (container: C, idx?: number, total?: number) => C | void

export interface NodeSolverDict<O> {
  html: (input: HTML) => O extends 'mirror' ? HTML : O
  text: (input: IText) => O extends 'mirror' ? IText : O
  element: (input: IElement, parent?: IElement | DocRoot) => O extends 'mirror' ? IElement : O
  node: (input: INode) => O extends 'mirror' ? INode : O
  document: (input: Document) => O extends 'mirror' ? Document : O
  fragment: (input: DocumentFragment) => O extends 'mirror' ? DocumentFragment : O
}

/**
 * A fully configured solver which is ready to convert a node into type `O`; if `O` is never
 * then it will mirror the input type as the output type
 */
export type NodeSolverReady<E extends NodeType, O> = <N extends Exclude<Container | HTML | null, E>>(node: N, parent?: IElement | DocRoot) => O extends 'mirror'
  ? N
  : O

export interface NodeSolverReceiver<E extends NodeType, O> {
  /** provide a solver dictionary */
  solver: (solver: Omit<NodeSolverDict<O>, E>) => NodeSolverReady<E, O>
}

export interface NodeSolverWithExclusions<E extends NodeType> {
  /** provide a type which all solvers will convert to */
  outputType: <O>() => NodeSolverReceiver<E, O>
  /** the input type should be maintained as the output type */
  mirror: () => NodeSolverReceiver<E, 'mirror'>
}

/**
 * Allows you to setup a type-string utility which receives DOM containers
 * and returns either the same container type or a specific type.
 *
 * This uses a builder API, of which this is the first step.
 */
export type NodeSolver = <E extends NodeType = never>(...exclude: E[]) => NodeSolverWithExclusions<E>

/**
 * A selector API provided by using the `select()` utility
 */
export interface NodeSelector<T extends Container | 'html'> {
  /**
   * The _type_ of the root node
   */
  type: () => NodeType
  /**
   * Find the first `IElement` found using the selector string.
   *
   * Note: _by default the return type includes `null` if no results were found but if you
   * prefer to throw an error you can state the error text and an error will be thrown_
   */
  findFirst: <E extends string | undefined>(sel: string, errorMsg?: E) => undefined extends E
    ? IElement | null
    : IElement
  /**
   * Find _all_ `IElement` results returned from the Selection query
   */
  findAll: <S extends string | undefined>(sel: S) => IElement[]

  /**
   * Allows the injection of a callback which will be used to mutate on the first `IElement` node
   * which matches the first
   */
  update: <E extends string | undefined>(sel?: string, errorMsg?: E) => (cb: UpdateCallback<IElement>) => NodeSelector<T>
  /**
   * Provides a way to inject an update callback which will be applied to all IElement nodes
   * which meet the selector query. If no query is provided, then this will be all `IElement`
   * children of the root node.
   */
  updateAll: <S extends string | undefined>(sel?: S) => (cb: UpdateCallback<IElement>) => NodeSelector<T>

  /**
   * Returns the root node with all mutations included
   */
  toContainer: () => undefined extends T
    ? DocumentFragment
    : T extends 'html'
      ? string
      : T
}

