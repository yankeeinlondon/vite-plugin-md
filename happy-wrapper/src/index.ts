export * from './attributes'
export * from './create'
export * from './diagnostics'
export * from './errors'
export * from './happy-types'
export * from './nodes'
export * from './select'
export * from './type-guards'
export * from './utils'
