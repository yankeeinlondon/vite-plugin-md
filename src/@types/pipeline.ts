import type MarkdownIt from 'markdown-it'
import type { UserConfig } from 'vite'
import type { EnumValues, Frontmatter, MetaProperty, ResolvedOptions } from './core'

export enum PipelineStage {
  /**
   * Initialized with incoming filename, config, options,
   * available events (core and provided by builders).
   */
  initialize = 'initialize',

  /**
   * All frontmatter has been extracted from default values and page values
   * but no mapping has been done yet.
   *
   * Note: this is the event hook which the included `meta` builder connects
   * to and it in turn _provides_ a `metaMapped` hook.
   */
  metaExtracted = 'metaExtracted',

  /**
   * The **MarkdownIt** parser is initialized, all builders
   * connecting at this point will receive a valid `md` parser
   * object so that they can participate in MD-to-HTML parsing.
   */
  parser = 'parser',
  /**
   * The **MarkdownIt** parser is initialized and all builders
   * have been able to apply their customizations to it.
   */
  parsed = 'parsed',

  /**
   * The `html` has now had all _script_ blocks extracted from it
   * and the script blocks are isolated in the `hoistedScripts`
   * property
   */
  hoisted = 'hoisted',

  /**
   * SFC blocks (template, script, and an array of customBlocks) are ready for
   * builders to inspect/mutate/etc.
   */
  sfcBlocksExtracted = 'sfcBlocksExtracted',

  /**
   * All mutations of page are complete; builders can hook into this stage but
   * will _not_ be able to mutate at this stage.
   */
  closeout = 'closeout',
}

export type IPipelineStage = EnumValues<PipelineStage>

export interface RulesUse {
  ruleName: string
  usage: 'adds' | 'patches' | 'modifies'
  description?: string
}

export type PipelineInitializer = (i?: Pipeline<PipelineStage.initialize>) => Pipeline<PipelineStage.initialize>

export interface BuilderRegistration<O extends BuilderOptions, H extends IPipelineStage > {
  name: string
  description?: string
  /** The lifecycle event/hook which this builder will respond to */
  lifecycle: H
  /**
   * The builder's handler function which receives the _payload_ for the
   * event lifecycle hook configured and then is allowed to mutate these
   * properties and pass back a similarly structured object to continue
   * on in that pipeline stage.
   */
  handler: BuilderHandler<O, H>

  /**
   * The options _specific_ to the builder
   */
  options: O

  /**
   * This isn't strictly required, but it is nice to express which rules you have used
   * modified, or added from the MarkdownIt parser.
   *
   * Note: builders should try to avoid mutating core rules; if they need a modification
   * for their purposes consider _monkey patching_ the rule so that downstream rules
   * have a better understanding of current rule state.
   */
  parserRules?: RulesUse[]

  /**
   * If this plugin needs to modify the configuration in some way at initialization
   * it can add a function here to do that. In most cases, the builder can simply
   * wait for their event hook to be called (at which point they will get the configuration
   * passed to them).
   */
  initializer?: BuilderHandler<O, PipelineStage.initialize>
}

/** container of events organized by PipelineStage */
export type BuilderStruct<T extends string> = Record<PipelineStage | T, any>

export interface AvailablePipelineProps {
  fileName: string
  /** the raw content in the file being processed */
  content: string
  /** the `vite-plugin-md` options */
  options: ResolvedOptions
  /** the Vite config */
  viteConfig: UserConfig

  /** the frontmatter metadata */
  frontmatter: Frontmatter
  /**
   * Meta properties that will be put into the HEAD section
   */
  meta: MetaProperty[]
  /**
   * Meta properties which are to be added to the VueJS router's "meta" attribute
   * for this page's route
   */
  routeMeta: Record<string, any>
  /**
   * Non-meta tags that will be put into the HEAD section of the page
   */
  head: Record<string, any>
  excerpt?: string
  /**
   * The markdown content (after extracting frontmatter)
   */
  md: string

  /** the **MarkdownIT** parser instance */
  parser: MarkdownIt

  /**
   * the HTML produced from MD content (and using parser rules passed in)
   */
  html: string

  /**
   * If any code blocks were found on the page then their languages will be represented
   * here.
   */
  fencedLanguages: Set<string>

  /**
   * all hoisted scripts
   */
  hoistedScripts: string[]

  /** the SFC's template block (aka, html content) */
  templateBlock: string
  /** the `<script [setup] ...>` block */
  scriptBlock: string
  /** any other top-level SFC blocks besides "template" and "script" */
  customBlocks: string[]

  /** the finalized component in string form */
  component: string
}

export type InitializedOmissions = 'md'
| 'fencedLanguages'
| 'frontmatter'
| 'head'
| 'meta'
| 'routeMeta'
| 'excerpt'
| 'html'
| 'hoistedScripts'
| 'templateBlock'
| 'parser'
| 'scriptBlock'
| 'customBlocks'
| 'component'

/** after extracting metadata */
export type MetaOmissions =
| 'fencedLanguages'
| 'parser'
| 'html'
| 'hoistedScripts'
| 'templateBlock'
| 'scriptBlock'
| 'customBlocks'
| 'component'

/** after providing the markdown-it parser */
export type ParserOmissions =
| 'fencedLanguages'
| 'html'
| 'hoistedScripts'
| 'templateBlock'
| 'scriptBlock'
| 'customBlocks'
| 'component'

/** after parsing to raw HTML using markdown-it */
export type ParsedOmissions =
| 'hoistedScripts'
| 'templateBlock'
| 'scriptBlock'
| 'customBlocks'
| 'component'

export type SfcBlockOmissions = 'component'

export type PipelineAvail<S extends IPipelineStage> = S extends 'initialize'
  ? Omit<AvailablePipelineProps, InitializedOmissions>
  : S extends 'metaExtracted' ? Omit<AvailablePipelineProps, MetaOmissions>
    : S extends 'parser' ? Omit<AvailablePipelineProps, ParserOmissions>
      : S extends 'parsed' ? Omit<AvailablePipelineProps, ParsedOmissions>
        : S extends 'sfcBlocksExtracted' ? Omit<AvailablePipelineProps, SfcBlockOmissions>
          : S extends 'closeout' ? AvailablePipelineProps
            : never

/**
 * The _state/payload_ that is available at a given stage in the pipeline process.
 *
 * - `<S>` provides the stage we're in
 * - `<E>` allows a builder to provide additional props for an event they are providing
 */
export type Pipeline<S extends IPipelineStage, E extends {} = {}> =
  PipelineAvail<S> & E

/**
 * The Builder's event listener/handler
 */
export type BuilderHandler<
  O extends BuilderOptions,
  E extends IPipelineStage,
  R extends Pipeline<E> = Pipeline<E>,
> = (payload: Pipeline<E>, options: O) => R

/**
 * Builder's must provide an export which meets this API constraint. Basic
 * structure of this higher order function is:
 *
 * - options() -> register( ) -> listen( ) -> payload
 */
export type BuilderApi<O extends {}, E extends IPipelineStage > = (options?: O) => () => BuilderRegistration<O, E>

export type InlineBuilder = <N extends string, L extends IPipelineStage>(name: N, lifecycle: L) => (payload: Pipeline<L>) => Pipeline<L>

/**
 * Builder options are expected to be a key/value dictionary but must
 * be allowed to be an empty object
 * */
export type BuilderOptions = Record<string, any> | {}

// TODO: BELOW is the type i'd like to use but can't find way to index the enum

// export type BuilderConfig = {
//   [K in keyof UnionToTuple<IPipelineStage>]: BuilderRegistration<any, PipelineStage[K]>[]
// }[keyof UnionToTuple<IPipelineStage>]

export type BuilderConfig = Record<IPipelineStage, BuilderRegistration<any, any>[] | []>
