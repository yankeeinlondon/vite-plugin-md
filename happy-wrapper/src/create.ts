import { identity } from 'fp-ts/lib/function'
import { Text, Window } from 'happy-dom'
import type { Document, DocumentFragment, IElement, IText } from 'happy-dom'
import { HappyMishap } from './errors'
import type { Container, HTML } from './happy-types'
import { isElement, isElementLike, isTextNodeLike } from './type-guards'
import { clone, solveForNodeType, toHtml } from './utils'

/**
 * Converts an HTML string into a Happy DOM document tree
 */
export function createDocument(body: string, head?: string): Document {
  const window = new Window()
  const document = window.document
  document.body.innerHTML = body
  if (head)
    document.head.innerHTML = head
  return document
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type FragmentFrom<_T extends Container | 'html'> = DocumentFragment

export function createFragment<C extends Container | HTML>(content?: C): FragmentFrom<C extends string ? 'html' : DocumentFragment> {
  const window = new Window()
  const document = window.document
  const fragment = document.createDocumentFragment() as DocumentFragment
  if (content)
    fragment.append(clone(content))

  return fragment as FragmentFrom<C extends string ? 'html' : DocumentFragment>
}

/**
 * Creates a DOM Node which will be either an `IElement` or `IText` node
 * based on the content passed in.
 */
export const createNode = (node: Container | string): IElement | IText => {
  const frag = createFragment(node)
  if (isElementLike(frag))
    return frag.firstElementChild as IElement
  else if (isTextNodeLike(frag))
    return frag.firstChild as IText
  else
    throw new HappyMishap('call to createNode() couldn\'t be converted to IElement or IText node', { name: 'createNode()', inspect: node })
}

/**
 * Creates a IText Node
 */
export function createTextNode(text?: string): IText {
  if (!text) {
    console.warn('An empty string was passed into createTextNode(); will be ignored but probably a mistake')
    return new Text('')
  }

  const frag = createFragment(text)
  if (isTextNodeLike(frag))
    return frag.firstChild as unknown as IText
  else
    throw new HappyMishap(`The HTML passed in cannot be converted to a single text node: "${text}".`, { name: 'createFragment(text)', inspect: frag })
}

/**
 * Creates an element node and can preserve parent relationship if known
 */
export const createElement = (el: Container | HTML, parent?: IElement): IElement => solveForNodeType()
  .outputType<IElement>()
  .solver({
    node: (n) => {
      if (isElement(n))
        return createElement(n) as IElement
      else
        throw new HappyMishap('can\'t create an IElement from an INode node because it doesn\'t have a tagName property', { inspect: n })
    },
    html: (h) => {
      const frag = createFragment(h)
      if (isElementLike(frag)) {
        if (parent) {
          parent.append(frag.firstElementChild)
          return parent?.lastElementChild
        }

        return frag.firstElementChild
      }
      else { throw new HappyMishap('The HTML passed into createElement() is not convertible to a IElement node!', { name: 'createElement(html)', inspect: frag }) }
    },
    element: identity,
    text: (t) => {
      throw new HappyMishap('An IElement can not be created from a IText node because element\'s require a wrapping tag name!', { name: 'createElement(text)', inspect: t })
    },
    fragment: (f) => {
      if (isElement(f.firstElementChild))
        return f.firstElementChild as IElement

      else
        throw new HappyMishap(`Unable to create a IElement node from: \n\n${toHtml(f)}`, { name: 'createElement()' })
    },
    document: (d) => {
      if (isElementLike(d)) {
        if (parent)
          throw new HappyMishap('A Document and a parent IElement were passed into createElement. This is not a valid combination!')

        return d.firstElementChild
      }

      else { throw new HappyMishap('Can not create an Element from passed in Document', { name: 'createElement(document)', inspect: d }) }
    },
  })(el)

export interface CssVariable {
  prop: string
  value: string
}

/**
 * Creates a new `<style>` node and provides a simple API surface to allow
 * populating the contents
 */
export const createInlineStyle = <T extends string>(type: T = 'text/css' as T) => {
  const node = createElement(`<style type="${type}" />`)

  const variables: Record<'root' | 'local', CssVariable[]> = {
    root: [],
    local: [],
  }

  const api = {
    addCssVariable(prop: string, value: string, scope: 'root' | 'local' = 'root') {
      if (!(scope in variables))
        variables[scope] = []

      variables[scope].push({ prop, value })

      return api
    },
    addClassDefinition(_klass: string, _value: string) {
      return api
    },

    addCssVariables(dictionary: Record<string, string>, scope: 'root' | 'local' = 'root') {
      Object.keys(dictionary).forEach(p => api.addCssVariable(p, dictionary[p], scope))

      return api
    },

    finish() {
      let text = ''
      // global variables
      Object.keys(variables.root).forEach((v, idx) => {
        text += idx === 0
          ? `:root {\n  --${v}: ${variables.root[v as any].value};`
          : `  --${v}: ${variables.root[v as any].value};`
      })
      if (variables.root.length > 0)
        text += '}\n'
      // local variables
      Object.keys(variables.local).forEach((v, idx) => {
        text += idx === 0
          ? `:root {\n  --${v}: ${variables.root[v as any].value};`
          : `  --${v}: ${variables.root[v as any].value};`
      })
      // properties

      node.append(text)
      return node
    },
  }

  return api
}
