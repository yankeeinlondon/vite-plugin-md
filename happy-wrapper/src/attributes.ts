import { pipe } from 'fp-ts/lib/function'
import type { IElement, INode } from 'happy-dom'
import { createFragment } from './create'
import { HappyMishap } from './errors'
import type { Container, DocRoot, GetAttribute, HTML } from './happy-types'
import { isDocument, isElement, isFragment } from './type-guards'
import { getNodeType, solveForNodeType, toHtml } from './utils'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttribute<T extends string> = (value: string) => <N extends Container | HTML>(node: N) => N

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type SetAttributeTo<_T extends string, _V extends string> = <N extends Container | HTML>(node: N) => N

export const setAttribute = <T extends string>(
  attr: T,
): SetAttribute<T> => <V extends string>(
    value: V,
  ): SetAttributeTo<T, V> => <N extends Container | HTML>(
      node: N,
    ): N => {
    const invalidNode = (n: INode) => {
      throw new HappyMishap(`You can not use the setAttribute() utility on a node of type: "${getNodeType(n)}"`, { name: `setAttribute(${attr})(${value})(INVALID)` })
    }
    const result = solveForNodeType()
      .mirror()
      .solver({
        html: h => pipe(h, createFragment, f => setAttribute(attr)(value)(f), toHtml),
        text: t => invalidNode(t),
        node: n => invalidNode(n),
        fragment: (f) => {
          f.firstElementChild.setAttribute(attr, value)
          return f
        },
        document: (d) => {
          d.body.firstElementChild.setAttribute(attr, value)
          return d
        },
        element: (e) => {
          e.setAttribute(attr, value)
          return e
        },
      })(node)

    return result
  }

export const getAttribute = <T extends string>(attr: T): GetAttribute<T> => {
  return solveForNodeType('text', 'node')
    .outputType<string>()
    .solver({
      html: h => pipe(h, createFragment, getAttribute(attr)),
      fragment: f => f.firstElementChild.getAttribute(attr),
      document: doc => doc.body.firstElementChild.getAttribute(attr),
      element: el => el.getAttribute(attr),
    })
}

const getClass = getAttribute('class')
const setClass = setAttribute('class')
/**
 * Provides the classes defined on a given container's top level
 * element as an array of strings
 */
export const getClassList = (container: Container | HTML | null): string[] => {
  if (!container)
    return []

  return solveForNodeType().outputType<string[]>().solver({
    html: h => pipe(h, createFragment, getClassList),
    document: d => getClass(d.body.firstElementChild)?.split(/\s+/) || [],
    fragment: f => getClass(f.firstElementChild)?.split(/\s+/) || [],
    element: e => getClass(e)?.split(/\s+/) || [],
    text: (n) => { throw new HappyMishap('Passed in a text node to getClassList!', { name: 'getClassList', inspect: n }) },
    node: (n) => { throw new HappyMishap('Passed in an unknown node type to getClassList!', { name: 'getClassList', inspect: n }) },
  })(container).filter(i => i)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type RemoveClass<R extends string | string[]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Removes a class from the top level node of a container's body.
 *
 * Note: if the class wasn't present then no change is performed
 */
export const removeClass = <R extends string | string[]>(
  remove: R,
): RemoveClass<R> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const current = getClass(doc)?.split(/\s+/g) || []
  const toRemove: string[] = !Array.isArray(remove) ? [remove] : remove

  const resultantClassString = Array.from(
    new Set<string>(current.filter(c => !toRemove.includes(c))),
  )
    .filter(i => i)
    .join(' ')

  return setClass(resultantClassString)(doc)
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type AddClass<A extends string[] | string[][]> = <C extends DocRoot | IElement | HTML>(container: C) => C

/**
 * Adds a class to the top level node of a document's body.
 */
export const addClass = <A extends string[] | string[][]>(
  ...add: A
): AddClass<A> => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const toAdd = (Array.isArray(add) ? add.flat() : [add]) as string[]

  const currentClasses = getClass(doc)?.split(/\s+/g) || []
  const resultantClasses = Array.from(new Set<string>([...currentClasses, ...toAdd]))

  return setClass(resultantClasses.join(' ').trim())(doc) as D
}

export type Filter = (string | RegExp)
/**
 * an array of filters but with an array passed in as first parameter, using
 * this data structure the initial array will be treated as "memory" for the
 * class items which were removed.
 */
export type FiltersWithMemory = [string[], ...Filter[]]

function isFiltersWithMemory(filters: Filter[] | FiltersWithMemory): filters is FiltersWithMemory {
  return Array.isArray(filters[0])
}

/**
 * Filters classes out to remove from a given element. As many filters as desired may be added where a
 * filter is:
 * - string - when a string it will compare for a direct match
 * - RegExp - will run the RegExp's `test(class)` method
 *
 * Optionally you may pass in a string array property as the first parameter and this will then
 * be populated with the filtered classes.
 */
export const filterClasses = <A extends Filter[] | FiltersWithMemory>(
  ...args: A
) => <D extends DocRoot | IElement | HTML>(doc: D): D => {
  const el = isDocument(doc) || isFragment(doc)
    ? doc.firstElementChild as IElement
    : isElement(doc)
      ? doc as IElement
      : null
  if (!el)
    throw new HappyMishap('An invalid container was passed into filterClasses()!', { name: 'filterClasses', inspect: doc })

  const filters = isFiltersWithMemory(args) ? args.slice(1) as Filter[] : args as Filter[]
  const memory: string[] = isFiltersWithMemory(args) ? args[0] : []
  const classes = getClassList(el)

  classes.forEach((klass) => {
    const matched = filters.every(f => typeof f === 'string'
      ? f === klass
      : f.test(klass),
    )
    if (matched)
      memory.push(klass)
  })

  return doc
}
