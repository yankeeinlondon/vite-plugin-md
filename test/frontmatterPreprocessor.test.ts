import { readFile } from 'fs/promises'
import { beforeAll, describe, expect, it } from 'vitest'
import { composeSfcBlocks } from '../src/pipeline'
import type { MetaProperty, ResolvedOptions } from '../src/@types'

const frontmatterPreprocess: ResolvedOptions['frontmatterPreprocess'] = (fm) => {
  const frontmatter = {
    title: 'default title',
    description: 'default description',
    ...fm,
  }
  const meta: MetaProperty[] = [
    { property: 'og:title', name: 'twitter:title', itemprop: 'title', content: frontmatter.title },
    {
      property: 'og:description',
      name: 'twitter:description',
      itemprop: 'description',
      content: frontmatter.description,
    },
  ]
  return {
    head: { ...frontmatter, meta },
    frontmatter: { ...frontmatter, meta },
    metaProps: [],
    routeMeta: {},
  }
}

let md = ''

describe('frontmatter pre-processor (without use of meta builder pattern)', () => {
  beforeAll(async() => {
    md = await readFile('test/fixtures/simple.md', 'utf-8')
  })

  it('frontmatter is unchanged', () => {
    const { meta } = composeSfcBlocks('', md, { frontmatterPreprocess })
    expect(meta.frontmatter).toMatchSnapshot()
  })

  it('head is unchanged', () => {
    const { meta } = composeSfcBlocks('', md, { frontmatterPreprocess })
    expect(meta.head).toMatchSnapshot()
  })

  it('meta props are unchanged', () => {
    const { meta } = composeSfcBlocks('', md, { frontmatterPreprocess })
    expect(meta.metaProps).toMatchSnapshot()
  })

  it('inline markdown is used over default properties', async() => {
    const { meta } = composeSfcBlocks('', md, { frontmatterPreprocess })

    // Positive tests
    expect(
      meta.frontmatter.title.includes('Hello World'),
      'the title attribute is retained over the default \'title\' value',
    ).toBeTruthy()

    expect(
      meta.frontmatter.description.includes('testing is the path to true happiness'),
      'description property is also retained',
    ).toBeTruthy()

    // Negative tests
    expect(
      meta.frontmatter.title.includes('default title'),
      'the title attribute is retained over the default \'title\' value',
    ).toBeFalsy()

    expect(meta.frontmatter.description.includes('default description'), 'default description is ignored').toBeFalsy()

    // Meta props
    const title = meta.head.meta.find(i => i.itemprop === 'title')
    const desc = meta.head.meta.find(i => i.itemprop === 'description')
    expect(title).toBeDefined()
    expect(desc).toBeDefined()
    expect(title?.property).toEqual('og:title')
    expect(desc?.property).toEqual('og:description')
  })
})
