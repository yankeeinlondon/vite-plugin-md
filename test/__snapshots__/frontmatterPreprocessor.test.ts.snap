// Vitest Snapshot v1

exports[`frontmatter pre-processor (without use of meta builder pattern) > frontmatter is unchanged 1`] = `
{
  "description": "testing is the path to true happiness",
  "meta": [
    {
      "content": "Hello World",
      "itemprop": "title",
      "name": "twitter:title",
      "property": "og:title",
    },
    {
      "content": "testing is the path to true happiness",
      "itemprop": "description",
      "name": "twitter:description",
      "property": "og:description",
    },
  ],
  "title": "Hello World",
}
`;

exports[`frontmatter pre-processor (without use of meta builder pattern) > head is unchanged 1`] = `
{
  "description": "testing is the path to true happiness",
  "meta": [
    {
      "content": "Hello World",
      "itemprop": "title",
      "name": "twitter:title",
      "property": "og:title",
    },
    {
      "content": "testing is the path to true happiness",
      "itemprop": "description",
      "name": "twitter:description",
      "property": "og:description",
    },
  ],
  "title": "Hello World",
}
`;

exports[`frontmatter pre-processor (without use of meta builder pattern) > meta props are unchanged 1`] = `[]`;
